Improve syscall implementation

wipawel opened the pull request:
Note: This PR is based on #282. Please review and merge #282 first or focus only on the new bits.

    Main change is a split of the syscall implementation with proper
    assembly routines: syscall_handler_entry() and simplified
    syscall_exit() and a simplified C-level syscall handler function.
    The syscall_exit() is implemented and handled fully in asm.
    
    The syscall() usermode function implements typical syscall ABI,
    passing syscall number via AX register and the rest of parameters
    via SI, DX, DI, R8 and R9. The assembly entry point rearranges the
    registers to fix the C-level handler function API.
    Only 5 arguments to the syscall can be provided.
    Usermode upon calling syscall() takes are of callee saved registers by
    saving/restoring them via usermode stack (only reminder of unused
    registers is preserved as the rest is considered implicitely clobbered).
    
    Additionally SYSCALL_PRINTF has been improved to support variadic
    number of arguments (it passes the va_list args to the kernel mode).

wipawel committed:
usermode: make explicit use of SYSCALL_EXIT

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
usermode: add USTR() macro to easily create user string literals

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
entry: add terminate_user_task() instead of using SYSCALL_EXIT

The terminate_user_task() is used upon exception when in usermode.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
asm-macros: add SYSRET and SWITCH_STACK macros

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
console: make vprintk() public

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
asm-macros: minor cleanup

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

wipawel committed:
usermode: improve syscall implementation

Main change is a split of the syscall implementation with proper
assembly routines: syscall_handler_entry() and simplified
syscall_exit() and a simplified C-level syscall handler function.
The syscall_exit() is implemented and handled fully in asm.

The syscall() usermode function implements typical syscall ABI,
passing syscall number via AX register and the rest of parameters
via SI, DX, DI, R8 and R9. The assembly entry point rearranges the
registers to fix the C-level handler function API.
Only 5 arguments to the syscall can be provided.
Usermode upon calling syscall() takes are of callee saved registers by
saving/restoring them via usermode stack (only reminder of unused
registers is preserved as the rest is considered implicitely clobbered).

Additionally SYSCALL_PRINTF has been improved to support variadic
number of arguments (it passes the va_list args to the kernel mode).

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

sktt commented on common/usermode.c:
```diff
@@ -177,67 +89,63 @@ static void init_syscall(void) {
+static inline long __user_text syscall(long syscall_nr, long arg1, long arg2, long arg3,
+                                       long arg4, long arg5) {
+    register long return_code asm(STR(_ASM_AX));
+    register long _arg4 asm("r8") = arg4;
+    register long _arg5 asm("r9") = arg5;
```
I thought arg4 and 5 would already be assigned r8 and r9?

wipawel replied:
They should be, yes - just wanted to be explicit and prevent reallocation.

sktt commented on common/usermode.c:
```diff
@@ -28,98 +28,19 @@
     case SYSCALL_EXIT:
-        syscall_restore();
-        _sys_exit();
         UNREACHABLE();
```
Maybe worth mentioning why it's unreachable.

wipawel replied:
Ok

sktt reviewed (dismissed):
lgtm, added a couple of remarks "just because".

minipli-oss commented on arch/x86/asm-offsets.c:
```diff
@@ -33,4 +33,37 @@
+    OFFSETOF(cpu_regs_r13, cpu_regs_t, r13);
+    OFFSETOF(cpu_regs_r14, cpu_regs_t, r14);
+    OFFSETOF(cpu_regs_r15, cpu_regs_t, r15);
+    OFFSETOF(cpu_regs_bx, cpu_regs_t, _ASM_BX);
```
duplicated line, already handled in line 46

minipli-oss commented on arch/x86/asm-offsets.c:
```diff
@@ -33,4 +33,37 @@
+    OFFSETOF(cpu_regs_r12, cpu_regs_t, r12);
+    OFFSETOF(cpu_regs_r13, cpu_regs_t, r13);
+    OFFSETOF(cpu_regs_r14, cpu_regs_t, r14);
+    OFFSETOF(cpu_regs_r15, cpu_regs_t, r15);
```
Should the 64 bit registers be wrapped in a `#ifdef __x86_64__` ... `#endif` block covering lines 52 to 59?
I mean, you already took care of making the code work for 32 bit by using the `_ASM_*` register name indirection. Would only make sense to guard the 64 bit registers as well.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -241,7 +241,21 @@ typedef uint64_t x86_reg_t;
 #endif
 
-struct cpu_regs {
+struct __packed cpu_exc {
```
Move `__packed` after the closing bracket, like:
```c
struct cpu_exc {
   ...
} __packed;
```

I see that we use both styles in KTF but the above seems to have more users and we should settle on one style only, imho.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -241,7 +241,21 @@ typedef uint64_t x86_reg_t;
+};
+typedef struct cpu_exc cpu_exc_t;
+
+struct __packed cpu_regs {
```
Dito, move the `__packed` attribute to the closing bracket.

minipli-oss commented on include/arch/x86/processor.h:
```diff
@@ -241,7 +241,21 @@ typedef uint64_t x86_reg_t;
 
-struct cpu_regs {
+struct __packed cpu_exc {
+    x86_ex_error_code_t error_code;
+    /* Populated by exception entry */
+    uint32_t vector;
```
As you're touching this code, can we please clarify the above "wrong" order? The CPU pushes an error code for certain exceptions on its own (#PF, #GP, #DF,...), so one might expect a different ordering but as these two members are 32 bit and the error code is handled as a 64 bit value, it's actually fine -- we simply make use of the upper 32 bits of the hw/sw generated error code and stuff in the vector number.

I'd suggest an anonymous union like this, maybe?:

```c
union {
    struct {
        x86_ex_error_code_t error_code;
        /* Populated by exception entry */
        uint32_t vector;
    };
    /* We stuff the vector number into the upper 32 bit of the error code */
    x86_reg_t _raw;
};
```

No need to mention this all break apart under 32 bit but it's already broken, so no need to worry ;)

wipawel replied:
This is handled by the exception_handler macro and the exception entry routines (entry_??) it generates.

The macro defines the entry_?? routines for all the exceptions like this:

```
exception_handler DE  X86_EX_DE  0
exception_handler DB  X86_EX_DB  0
exception_handler NMI X86_EX_NMI 0
exception_handler BP  X86_EX_BP  0
exception_handler OF  X86_EX_OF  0
exception_handler BR  X86_EX_BR  0
exception_handler UD  X86_EX_UD  0
exception_handler NM  X86_EX_NM  0
exception_handler DF  X86_EX_DF  1
exception_handler CS  X86_EX_CS  0
exception_handler TS  X86_EX_TS  1
exception_handler NP  X86_EX_NP  1
exception_handler SS  X86_EX_SS  1
exception_handler GP  X86_EX_GP  1
exception_handler PF  X86_EX_PF  1
exception_handler SPV X86_EX_SPV 0
exception_handler MF  X86_EX_MF  0
exception_handler AC  X86_EX_AC  1
exception_handler MC  X86_EX_MC  0
exception_handler XM  X86_EX_XM  0
exception_handler VE  X86_EX_VE  0
exception_handler SE  X86_EX_SE  1
```

and the macro implementation:
```
.macro exception_handler sym vec has_error_code
ENTRY(entry_\sym)
    enter_from_usermode

    .if \has_error_code == 0
        push $0
    .endif

    movl $\vec, cpu_exc_vector(%_ASM_SP)
    jmp handle_exception
END_FUNC(entry_\sym)
.endm
````

First parameter is the last part of the entry_?? name (the ??).
Second parameter is the actual exception vector.
The third parameter is an indicator whether the exception pushes error_code or not.

If the exception does not push the error_code the entry routine does so anyway (it pushes 0 then) to have the stack layout unified.

But, yeah it was broken for 32bit mode.

I will fix it with the following change:

```
--- a/arch/x86/entry.S
+++ b/arch/x86/entry.S
@@ -68,7 +68,12 @@ ENTRY(entry_\sym)
         push $0
     .endif
 
+#if defined(__x86_64__)
     movl $\vec, cpu_exc_vector(%_ASM_SP)
+#else
+    push $\vec
+#endif
+
     jmp handle_exception
 END_FUNC(entry_\sym)
 .endm
@@ -119,11 +124,7 @@ ENTRY(handle_exception)
 
     RESTORE_ALL_REGS
 
-#if defined(__x86_64__)
     add $8, %_ASM_SP
-#else
-    add $4, %_ASM_SP
-#endif
 
     enter_to_usermode
     IRET
```

minipli-oss replied:
Unfortunately, that's not sufficient, as `struct cpu_exc` just doesn't fit the 32 bit exception frame (excessive padding for the segment registers, still mixed up ordering for `error_code` and `vector`). But that shouldn't hold the current PR.

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -153,6 +154,16 @@ ENTRY(enter_usermode)
+    movl $-EFAULT, cpu_regs_ax(%_ASM_SP)
+    RESTORE_ALL_REGS
+
+    ret
```
Uhm, this function is not supposed to return via `RET`, right? It restores (read: clobbers) **all** registers, thereby breaks **any** call convention guarantees. Or is that intentional? What's that function supposed to do?

wipawel replied:
Yes, this is intentional. We are already in the kernel, but because the exception happened in usermode we kept the user stack, displayed the exception in `do_exception()` and jumped to this function. Now the function switches the stack to kernel one, and undoes what `enter_usermode()` did (pops flags, saves return code into `AX`, which is -EFAULT for exceptions in usermode and restores previously saved kernel regs). The `RET` returns to the caller of `enter_usermode()`.
Which is the user task creation function.

minipli-oss replied:
Ahh, thanks for the explanation!

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -154,6 +166,16 @@ ENTRY(enter_usermode)
+    mov %_ASM_SI, cpu_regs_ax(%_ASM_SP)
+    RESTORE_ALL_REGS
+
+    ret
```
Related to my confusion about `terminate_user_task`, where does this `RET` lead to? It's clearly intended to end up in kernel code, as we're running with kernel GS and CR3. But I fail to see where, clearly lacking some context here.

wipawel replied:
The RET returns to the caller of `enter_usermode()`, see the previous comment.

minipli-oss commented on arch/x86/entry.S:
```diff
@@ -164,6 +186,28 @@ ENTRY(terminate_user_task)
+
+    syscall_to_usermode
+
+    SYSRET
```
I don't know what guarantees KTF wants to provide for the syscall ABI. But right now it clobbers all the caller-saved registers (RAX (intentional, as return value register), RDX, RSI, R8-R10). Or, to read it differently, it _leaks_ those registers to userland.

We might want to do a follow up patch that sanitizes the callee-clobbered registers to:
1. avoid speculation based on user provided register values and
2. prevent leaking kernel data to userland / clobber user register state.

wipawel replied:
Modulo the callee saved registers handled in kernel instead of user mode, I am not sure we care much about the security aspects. After all, we always run trusted code that does malicious things in KTF :-)

minipli-oss replied:
I was just concerned about getting test flawed results by seeing leaked kernel registers.

minipli-oss commented on common/usermode.c:
```diff
@@ -177,67 +89,63 @@ static void init_syscall(void) {
+        SAVE_CALLEE_SAVED_REGS()
+        "syscall\n"
+        RESTORE_CALLEE_SAVED_REGS()
+        : "=a"(return_code)
```
Ahh, I see. You worked around the register clobbering by saving/restoring in userland. But to get the privilege boundaries right, that should be done in the kernel.

wipawel replied:
Ok, as Johannes also suggested doing this in the kernel, I will change accordingly.

minipli-oss reviewed (changes requested):
Great series, thanks a lot for doing this! Only needs a tiny bit more polishing.

wipawel committed:
arch,entry: fix error_code and vector storing for 32bit mode

In 64bit mode push 64bit 0 value on the stack as the aligning error
code and overwrite upper 32 bits of it with vector value.

In 32bit mode push 32bit 0 value on the stack as the aligning error
code and push vector value additionally.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@grsecurity.net>

minipli-oss reviewed (approved):
LGTM!
